parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+            return Output()
+
+        cands, cand_vecs, label_inds = self._build_candidates(
+            batch, source=self.opt['candidates'], mode='train')
+
+        if self.opt.get('input_dropout', 0) > 0:
+            batch_text, batch_label, batch_cand_vecs = self.input_dropout(
+                batch.text_vec,
+                batch.label_vec,
+                cand_vecs
+            )
+            xe, ye = self.model(
+                batch_text,
+                ys=batch_label,
+                cands=batch_cand_vecs
+            )
+        else:
+            xe, ye = self.model(
+                batch.text_vec,
+                ys=batch.label_vec,
+                cands=cand_vecs
+            )
+        xe = torch.cat([xe for _ in range(ye.size(1))], dim=1)
+
+        if self.use_cuda:
+            y = -torch.ones(xe.size(0), xe.size(1)).cuda()
+            y_0 = torch.ones(xe.size(0)).cuda()
+        else:
+            y = -torch.ones(xe.size(0), xe.size(1))
+            y_0 = torch.ones(xe.size(0))
+        y[:, 0] = y_0
+
+        loss = self.criterion(
+            xe.view(-1, xe.size(-1)),
+            ye.view(-1, xe.size(-1)),
+            y.view(-1)
+        )
+        loss.backward()
+        self.optimizer.step()
+        scores = nn.CosineSimilarity(dim=-1).forward(xe, ye)
+
+        # Update metrics
+        self.metrics['loss'] += loss.item()
+        self.metrics['examples'] += batchsize
+        _, ranks = scores.sort(1, descending=True)
+        for b in range(batchsize):
+            rank = (ranks[b] == label_inds[b]).nonzero().item()
+            self.metrics['rank'] += 1 + rank
+
+        return Output()
+
+    def eval_step(self, batch):
+        """Evaluate a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.eval()
+
+        cands, cand_vecs, label_inds = self._build_candidates(
+            batch, source=self.opt['eval_candidates'], mode='eval')
+
+        xe, ye = self.model(
+            batch.text_vec,
+            ys=None,
+            cands=cand_vecs
+        )
+        xe = torch.cat([xe for _ in range(ye.size(1))], dim=1)
+
+        scores = nn.CosineSimilarity(dim=-1).forward(xe, ye)
+        _, ranks = scores.sort(1, descending=True)
+
+        self.metrics['examples'] += batchsize
+        for b in range(batchsize):
+            rank = (ranks[b] == label_inds[b]).nonzero().item()
+            self.metrics['rank'] += 1 + rank
+
+        cand_preds = []
+        for i, ordering in enumerate(ranks):
+            if cand_vecs.dim() == 2:
+                cand_list = cands
+            elif cand_vecs.dim() == 3:
+                cand_list = cands[i]
+            cand_preds.append([cand_list[rank] for rank in ordering])
+        preds = [cand_preds[i][0] for i in range(batchsize)]
+        return Output(preds, cand_preds)
+
+    def observe(self, observation):
+        labels = observation.get('labels')
```

Comment:
This part could use some comments, because it's not obvious what's going on to me.